aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,349 @@
+            read_curve: copy byte_curve,
+            create_curve: copy byte_curve,
+            write_curve: copy byte_curve,
+        };
+        validate_usage_config(&item_config);
+        validate_usage_config(&item_config);
+        move_to(aptos_framework, StorageGasConfig {
+            item_config,
+            byte_config,
+        });
+
+        assert!(
+            !exists<StorageGas>(@aptos_framework),
+            error::already_exists(ESTORAGE_GAS)
+        );
+        move_to(aptos_framework, StorageGas {
+            per_item_read: 0,
+            per_item_create: 0,
+            per_item_write: 0,
+            per_byte_read: 0,
+            per_byte_create: 0,
+            per_byte_write: 0,
+        });
+    }
+
+    fun validate_usage_config(config: &UsageGasConfig) {
+        assert!(config.target_usage > 0, error::invalid_argument(EZERO_TARGET_USAGE));
+        assert!(config.target_usage <= 18446744073709551615 / BASIS_POINT_DENOMINATION, error::invalid_argument(ETARGET_USAGE_TOO_BIG));
+        validate_curve(&config.read_curve);
+        validate_curve(&config.create_curve);
+        validate_curve(&config.write_curve);
+    }
+
+    fun validate_curve(curve: &GasCurve) {
+        assert!(curve.max_gas >= curve.min_gas, error::invalid_argument(EINVALID_GAS_RANGE));
+        let points = &curve.points;
+        let len = vector::length(points);
+        assert!(len >= 2, error::invalid_argument(EINVALID_CURVE_LENGTH));
+        assert!(vector::borrow(points, 0).x > 0, error::invalid_argument(EINVALID_CURVE_START_POINT));
+        assert!(vector::borrow(points, len - 1).x < BASIS_POINT_DENOMINATION, error::invalid_argument(EINVALID_CURVE_END_POINT));
+        let i = 0;
+        while (i <= len) {
+            let cur = if (i == 0) { &Point {x: 0, y: 0} } else { vector::borrow(points, i - 1) };
+            let next = if (i == len) { &Point {x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION} } else { vector::borrow(points, i) };
+            assert!(cur.x < next.x && cur.y <= next.y, error::invalid_argument(EINVALID_MONOTONICALLY_NON_DECREASING_CURVE));
+            i = i + 1;
+        }
+    }
+
+    fun calculate_gas(max_usage: u64, current_usage: u64, curve: &GasCurve): u64 {
+        let capped_current_usage = if (current_usage > max_usage) max_usage else current_usage;
+        let points = &curve.points;
+        let num_points = vector::length(points);
+        let current_usage_bps = capped_current_usage * BASIS_POINT_DENOMINATION / max_usage;
+        let (i, j) = (0, num_points - 1);
+        while (i < j) {
+            let mid = j - (j - i) / 2;
+            if (current_usage_bps < vector::borrow(points, mid).x) {
+                j = mid - 1;
+            } else {
+                i = mid;
+            };
+        };
+        // Check the corner case that current_usage_bps drops before the first point.
+        let (left, right) = if (current_usage_bps < vector::borrow(points, i).x) {
```

Comment:
This could only happen when `i == 0`, right?

Reply:
yes. it is a corner case.

Reply:
Then I wonder if it's clearer to replace the `i`s with `0`s?